pkg/collector/containerlogs_containerd_collector.go

```diff
@@ -0,0 +1,160 @@
+package collector
+
+import (
+	"os"
+	"path"
+	"strings"
+
+	"github.com/Azure/aks-periscope/pkg/utils"
+)
+
+// ContainerLogsCollectorContainerD defines a ContainerLogs Collector struct for containerd clusters
+type ContainerLogsContainerDCollector struct {
+	data map[string]string
+}
+
+// NewContainerLogsCollectorContainerD is a constructor
+func NewContainerLogsContainerDCollector() *ContainerLogsContainerDCollector {
+	return &ContainerLogsContainerDCollector{
+		data: make(map[string]string),
+	}
+}
+
+func (collector *ContainerLogsContainerDCollector) GetName() string {
+	return "containerlogscontainerd"
+}
+
+type ContainerLog struct {
+	podname       string
+	namespace     string
+	containerName string
+	containeruid  string
+	filepath      string
+}
+
+type ContainerLogSelector struct {
+	namespace           string
+	containerNamePrefix string
+}
+
+//special selector which indicates logs from containers in all namespaces should be selected for collection
+const allNamespacesSelector = "--all-namespaces"
+
+// NOTE pod log files not currently used are in sub-directories of /var/log/pods and have format NAMESPACE_PODNAME_SOMEGUIDMEANINGTBD/CONTAINERNAME/#.log,
+// where I assume # is incremented and a new log created for each container restart (starts at # == 0)
+const containerLogDirectory = "/var/log/containers"
+
+func (collector *ContainerLogsContainerDCollector) GetData() map[string]string {
+	return collector.data
+}
+
+// Collect implements the interface method
+func (collector *ContainerLogsContainerDCollector) Collect() error {
+	selectorStrings := strings.Fields(os.Getenv("DIAGNOSTIC_CONTAINERLOGS_LIST"))
+
+	containerLogSelectors := parseContainerLogSelectors(selectorStrings)
+
+	allContainersEverRun, err := getAllContainerLogFilesThatHaveEverRunOnHost()
+	if err != nil {
+		return err
+	}
+
+	containerLogs := parseContainerLogFilenames(containerLogDirectory, allContainersEverRun)
+
+	containerLogsToCollect := determineContainerLogsToCollect(containerLogs, containerLogSelectors)
+
+	for _, containerLog := range containerLogsToCollect {
+		output, err := utils.RunCommandOnHost("cat", containerLog.filepath)
+		if err != nil {
+			return err
+		}
+
+		collector.data[containerLog.containerName] = output
+	}
+
+	return nil
+}
+
+func determineContainerLogsToCollect(allContainers []ContainerLog, selectors []ContainerLogSelector) []ContainerLog {
+	var selectedContainerLogs []ContainerLog
+	for _, containerLog := range allContainers {
+		for _, selector := range selectors {
+			if doesSelectorSelectContainerLog(containerLog, selector) {
+				selectedContainerLogs = append(selectedContainerLogs, containerLog)
+			}
+		}
+	}
+	return selectedContainerLogs
+}
+
+func doesSelectorSelectContainerLog(containerLog ContainerLog, selector ContainerLogSelector) bool {
+	if selector.namespace == allNamespacesSelector {
```

Comment:
why don't u put all conditions in one if?

return selector.namespace == allNamespacesSelector && containerLog.namespace == selector.namespace && strings.HasPrefix(containerLog.containerName, selector.containerNamePrefix)

Reply:
The line you pasted would change the functionality, an equivalent single line would be:
return (selector.namespace == allNamespacesSelector) || (containerLog.namespace == selector.namespace && strings.HasPrefix(containerLog.containerName, selector.containerNamePrefix))

But the current form keeps the "special behaviour" as a separate case and is more readable